### PR TITLE
Enable GTP for heterogeneous MIMO training

### PR DESCRIPTION
**Pull request by yashaswikarnati**
## Summary

- extend heterogeneous MIMO grids with dense GTP and expert GTP axes and thread their process groups through data distribution, bridge communication, RNG, gradient synchronization, optimizer ownership, and checkpointing
- pass `ProcessGroupCollection` through Mamba projections, shared experts, language embeddings, and the output head
- distinguish full data-distribution groups from dense/expert replica groups in distributed checkpoint metadata and support fresh MIMO hybrid checkpoint restore
- parameterize the 20-layer Nemotron mock launcher for multi-node GTP/EGTP execution

## Why

The heterogeneous MIMO path builds disjoint module grids instead of using MPU-global groups. On top of #6234, it still lacked the GTP/EGTP axes and process-group plumbing required by the dense, expert, bridge, optimizer, and distributed-checkpoint paths.

Depends on #6234.

## Validation

- 24-rank TP2/GTP8/EP4/EGTP2, 128-expert, 20-layer training completed 100 iterations with finite loss and gradient norm, `skipped=0`, and `nan=0`
- saved a distributed checkpoint at iteration 100; a fresh 24-rank process loaded the model and distributed Adam state and completed a finite iteration 101
- focused unit tests: 15 passed; distributed topology/bridge: 2 passed; TE GTP DCP: 16 passed, 2 skipped
- Ruff, isort, Python compilation, shell syntax, and diff checks passed
- runtime validation used Transformer Engine 2.19.dev0 or newer, as required by #6234

### CODE REVIEW

File: docs/api-guide/core/generalized_tensor_parallel.md [hunk 1]
```diff
@@ -592,12 +552,11 @@ torchrun --nproc-per-node 4 -m pytest tests/unit_tests/generalized_tensor_parall
...
 | `test_gtp_cudagraph_grad.py` | Capture-step grad-norm guard (§1.2): `_backup_grads_before_capture`/`_restore_grads_after_capture` keep a graph capture from clobbering finalized `main_grad` (own params + cross-graph `next_w`, incl. routed-expert `weight_list`). |
 | `test_gtp_dcp.py` | DCP sharding metadata (§3.3): TP×GTP_remat offsets, pad reshard, `replica_id`, native-FP8 save/load. |
 | `test_gtp_muon_dcp.py` | Muon optimizer-state DCP roundtrip (§1.6): `replica_id` fold + native-FP8 backfill matching. |
 | `test_gtp_fp8_param_gather.py` | Native-FP8 GTP_remat (§1.3): fp8-vs-BF16 loss parity (TP1/TP2, MoE), post-save-spike guard. |
-
-The fp32-accumulation primitive itself is covered outside this suite, by `tests/unit_tests/distributed/test_reduce_scatter_with_fp32_accumulation.py`, which does not require GTP_remat.
+| `test_gtp_custom_pgs.py` | `pg_collection` plumbing: a custom `gtp_remat` group (permuted ranks, same size) must give the same fwd/bwd results as the MPU groups — catches modules reading `parallel_state` instead of the collection passed to them. |
```
Comment (yashaswikarnati) [inline docs/api-guide/core/generalized_tensor_parallel.md:560]: remove

File: examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh [hunk 2]
```diff
@@ -1,26 +1,78 @@
 #!/bin/bash
 
-# Run an eight-rank heterogeneous mock training loop with Nemotron6-MoE VLM 20L.
+# Run heterogeneous mock training with the Nemotron6-MoE VLM 20L recipe.
 
 set -euo pipefail
 
+export CUDA_DEVICE_MAX_CONNECTIONS=${CUDA_DEVICE_MAX_CONNECTIONS:-1}
```
Comment (yashaswikarnati) [inline examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh:7]: why did we add this back ? not needed ?

File: examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh [hunk 3]
```diff
@@ -1,26 +1,78 @@
 #!/bin/bash
 
-# Run an eight-rank heterogeneous mock training loop with Nemotron6-MoE VLM 20L.
+# Run heterogeneous mock training with the Nemotron6-MoE VLM 20L recipe.
```
Comment (yashaswikarnati) [inline examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh:3]: no need to change this comment, avoid unnecessary edits

File: examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh [hunk 4]
```diff
@@ -1,26 +1,78 @@
...
+LLM_CP=${LLM_CP:-1}
+LLM_PP=${LLM_PP:-1}
+LLM_DP=${LLM_DP:-2}
+LLM_EP=${LLM_EP:-4}
+LLM_EXPT_TP=${LLM_EXPT_TP:-1}
+TENSOR_PARALLEL_NUM_WEIGHT_SHARDS=${TENSOR_PARALLEL_NUM_WEIGHT_SHARDS:-${LLM_TP}}
+EXPERT_TENSOR_PARALLEL_NUM_WEIGHT_SHARDS=${EXPERT_TENSOR_PARALLEL_NUM_WEIGHT_SHARDS:-${LLM_EXPT_TP}}
+
+if ((TENSOR_PARALLEL_NUM_WEIGHT_SHARDS % LLM_TP != 0)); then
+  echo "TENSOR_PARALLEL_NUM_WEIGHT_SHARDS must be divisible by LLM_TP" >&2
```
Comment (yashaswikarnati) [inline examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh:29]: avoid these verbose checks in the sbatch script, also dont introduce new env variables if not necessary

File: examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh [hunk 5]
```diff
@@ -1,26 +1,78 @@
...
+fi
+
+GLOBAL_BATCH_SIZE=${GLOBAL_BATCH_SIZE:-$((MICRO_BATCH_SIZE * NUM_MICROBATCHES * LLM_DP * GTP))}
 TORCHRUN_LOG_DIR=${TORCHRUN_LOG_DIR:-"${PWD}/logs/torchrun-$(date +%Y%m%d_%H%M%S)-$$"}
 mkdir -p "${TORCHRUN_LOG_DIR}"
 
 TORCHRUN_ARGS=(
-  --standalone
-  --nproc-per-node 8
+  --nproc-per-node "${NPROC_PER_NODE}"
```
Comment (yashaswikarnati) [inline examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh:51]: the diff should be minimal dont change unnecessary things

File: examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh [hunk 6]
```diff
@@ -71,15 +123,17 @@ uv run --extra ssm python -m torch.distributed.run \
   --seq-length 8192 \
   --max-position-embeddings 8192 \
   --bf16 \
-  --encoder-tp 2 \
-  --encoder-dp 2 \
-  --llm-offset 4 \
-  --llm-tp 2 \
-  --llm-cp 1 \
-  --llm-pp 1 \
+  --encoder-tp "${ENCODER_TP}" \
```
Comment (yashaswikarnati) [inline examples/mimo/scripts/run_hetero_nemotron_20l_mock_train.sh:126]: we dont need env variables for this

File: examples/mimo/pretrain_mimo.py [hunk 7]
```diff
@@ -48,14 +48,19 @@ def _parse_and_validate() -> argparse.Namespace:
     args = parse_args(extra_args_provider)
     validate_hetero_grid_args(args, args.world_size)
     physical_world_size = args.world_size
-    # Stock validate_args sets data_parallel_size = world_size // (tp*pp*cp); feed the
-    # language module's world (llm_dp; stock tp/pp/cp stay 1, MIMO parallelism is in --llm-*)
-    # so it yields llm_dp. The physical world incl. encoder ranks is restored below.
+    # Stock validation owns the derived training arguments. Give it the language module's
+    # parallel degrees and logical world so its DP/GTP accounting matches the explicit MIMO grid.
+    args.tensor_model_parallel_size = args.llm_tp
```
Comment (yashaswikarnati) [inline examples/mimo/pretrain_mimo.py:53]: why did we newly introduce this   args.tensor_model_parallel_size = args.llm_tp
    args.pipeline_model_parallel_size = args.llm_pp
    args.context_parallel_size = args.llm_cp
    args.expert_model_parallel_size = args.llm_ep
    args.expert_tensor_parallel_size = args.llm_expt_tp or 1

File: megatron/core/distributed/finalize_model_grads.py [hunk 8]
```diff
@@ -604,12 +601,29 @@ def finalize_model_grads(
...
         dp_cp_group = getattr(pg_collection, 'dp_cp_gtp_remat', None) or pg_collection.dp_cp
+        gtp_remat_group = getattr(pg_collection, 'gtp_remat', None)
+        egtp_remat_group = getattr(pg_collection, 'expt_gtp_remat', None)
     else:
         tp_group = parallel_state.get_tensor_model_parallel_group()
         pp_group = parallel_state.get_pipeline_model_parallel_group()
         embd_group = parallel_state.get_embedding_group(check_initialized=False)
         pos_emb_group = parallel_state.get_position_embedding_group(check_initialized=False)
         dp_cp_group = parallel_state.get_data_parallel_group(with_context_parallel=True)
+        gtp_remat_group = parallel_state.get_gtp_weight_remat_group(check_initialized=False)
```
Comment (yashaswikarnati) [inline megatron/core/distributed/finalize_model_grads.py:612]: is these changes already on main ?

File: megatron/core/pipeline_parallel/bridge_communicator.py [hunk 9]
```diff
@@ -14,11 +14,11 @@
 class CommRole(Enum):
     """Communication role for ranks in bridge communication.
 
-    SENDER: Leader tp-cp rank within each DP replica of source grid.
+    SENDER: Leader tp-cp rank within each data lane of source grid.
```
Comment (yashaswikarnati) [inline megatron/core/pipeline_parallel/bridge_communicator.py:17]: this change should already be in main ? for bridge comm? why we need these changes again here ? seems only cosmetic ? https://github.com/NVIDIA/Megatron-LM/pull/6263

File: megatron/core/pipeline_parallel/schedules.py [hunk 10]
```diff
@@ -690,6 +690,15 @@ def _build_default_pg_collection() -> ProcessGroupCollection:
     pg_collection.dp = parallel_state.get_data_parallel_group(
         with_context_parallel=False, partial_data_parallel=False
     )
+    # gtp_remat axis: consumers read these with getattr and silently skip the gtp_remat
```
Comment (yashaswikarnati) [inline megatron/core/pipeline_parallel/schedules.py:693]: this change already seems to be on main

File: megatron/core/pipeline_parallel/schedules.py [hunk 11]
```diff
@@ -1614,7 +1623,7 @@ def forward_backward_helper_wrapper(
                 recv_next = True
                 if is_pp_last_stage(p2p_communicator.pp_group):
                     recv_next = False
-                (input_tensor, output_tensor_grad) = (
+                input_tensor, output_tensor_grad = (
```
Comment (yashaswikarnati) [inline megatron/core/pipeline_parallel/schedules.py:1626]: cosmetic changes and already on main not necessary

File: megatron/core/tensor_parallel/generalized_tensor_parallelism.py [hunk 12]
```diff
@@ -2255,6 +2149,7 @@ def make_sharded_tensors_for_checkpoint_with_gtp_remat(
     tp_group,
     dp_cp_group,
     intra_dp_cp_group=None,
+    intra_expt_dp_group=None,
```
Comment (yashaswikarnati) [inline megatron/core/tensor_parallel/generalized_tensor_parallelism.py:2152]: why do we need changes to this file ? seems not relevant ?

File: megatron/core/tensor_parallel/inference_layers.py [hunk 13]
```diff
@@ -139,6 +143,7 @@ def __init__(
         tp_comm_buffer_name: Optional[str] = None,
         tp_group: Optional[torch.distributed.ProcessGroup] = None,
         name: str | None = None,
+        pg_collection: Optional[ProcessGroupCollection] = None,
```
Comment (yashaswikarnati) [inline megatron/core/tensor_parallel/inference_layers.py:146]: inference layers changes is out of scope

File: megatron/core/transformer/attention.py [hunk 14]
```diff
@@ -421,6 +421,7 @@ def __init__(
             is_expert=False,
             tp_comm_buffer_name='proj',
             tp_group=self.pg_collection.tp,
+            pg_collection=self.pg_collection,
```
Comment (yashaswikarnati) [inline megatron/core/transformer/attention.py:424]: are these changes not on main already ?
